optuna/integration/pytorch_distributed.py

```diff
@@ -0,0 +1,249 @@
+        dist.barrier()
+
+    def set_system_attr(self, key: str, value: Any) -> None:
+
+        if dist.get_rank() == 0:
+            assert self.delegate is not None
+            self.delegate.set_system_attr(key, value)
+        dist.barrier()
+
+    @property
+    def number(self) -> int:
+        def func() -> int:
+
+            assert self.delegate is not None
+            return self.delegate.number
+
+        return _call_and_communicate(func, torch.int)
+
+    @property
+    def trial_id(self) -> int:
+
+        warnings.warn(
+            "The use of `TorchDistributedTrial.trial_id` is deprecated. "
+            "Please use `TorchDistributedTrial.number` instead.",
+            FutureWarning,
+        )
+        return self._trial_id
+
+    @property
+    def _trial_id(self) -> int:
+        def func() -> int:
+
+            assert self.delegate is not None
+            return self.delegate._trial_id
+
+        return _call_and_communicate(func, torch.int)
+
+    @property
+    def params(self) -> Dict[str, Any]:
+        def func() -> Dict[str, Any]:
+
+            assert self.delegate is not None
+            return self.delegate.params
+
+        return _call_and_communicate_obj(func)
+
+    @property
+    def distributions(self) -> Dict[str, BaseDistribution]:
+        def func() -> Dict[str, BaseDistribution]:
+
+            assert self.delegate is not None
+            return self.delegate.distributions
+
+        return _call_and_communicate_obj(func)
+
+    @property
+    def user_attrs(self) -> Dict[str, Any]:
+        def func() -> Dict[str, Any]:
+
+            assert self.delegate is not None
+            return self.delegate.user_attrs
+
+        return _call_and_communicate_obj(func)
+
+    @property
+    def system_attrs(self) -> Dict[str, Any]:
+        def func() -> Dict[str, Any]:
+
+            assert self.delegate is not None
+            return self.delegate.system_attrs
+
+        return _call_and_communicate_obj(func)
+
+    @property
+    def datetime_start(self) -> Optional[datetime]:
+        def func() -> Optional[datetime]:
+
+            assert self.delegate is not None
+            return self.delegate.datetime_start
+
+        return _call_and_communicate_obj(func)
+
+
+def _call_and_communicate(func: Callable, dtype: "torch.dtype") -> Any:
+    buffer = torch.empty(1, dtype=dtype)
+    rank = dist.get_rank()
+    if rank == 0:
+        result = func()
+        buffer[0] = result
+    if dist.get_backend() == "nccl":
+        buffer = buffer.cuda(torch.device(rank))
+    dist.broadcast(buffer, src=0)
+    return buffer.item()
+
+
+def _call_and_communicate_obj(func: Callable) -> Any:
+    buffer = None
+    size_buffer = torch.empty(1, dtype=torch.int)
+    rank = dist.get_rank()
+    if rank == 0:
+        result = func()
+        buffer = _to_tensor(result)
+        size_buffer[0] = buffer.shape[0]
+    if dist.get_backend() == "nccl":
+        size_buffer = size_buffer.cuda(torch.device(rank))
```

Comment:
Cann't we just call `buffer.cuda()` without the argument `torch.device(rank)`?

I saw that [the PyTorch official doc](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.cuda) says the default value of the first argument is "the current CUDA device".
(This is a curious question as I am not familiar with PyTorch)

Reply:
Thank you for your question.
In my environment, processes stalled if I didn't specify the rank to `.cuda`. I guess DDP assumes that each process uses its own device, and the device for the communication is required to be set explicitly.

On the other hand, I saw the following warning message when I removed the `device_ids` argument from `DDP.__init__`. It may suggest we can remove the rank from `.cuda()` if we set `CUDA_VISIBLE_DEVICES`. For example, we can set `CUDA_VISIBLE_DEVICE=RANK`. In this case, setting the rank to `.cuda` may cause a problem because only the device 0 is visible to each process.

> Single-Process Multi-GPU is not the recommended mode for DDP. In this mode, each DDP instance operates on multiple devices and creates multiple module replicas within one process. The overhead of scatter/gather and GIL contention in every forward pass can slow down training. Please consider using one DDP instance per device or per module replica by explicitly setting device_ids or CUDA_VISIBLE_DEVICES.

I'll investigate it in detail.

Reply:
I investigated this issue and found two problems to set `rank` to the device.  To resolve these problems, I think the device should be selected by users, and I added the `device` argument to `TorchDistributedTrial` in the commit https://github.com/optuna/optuna/pull/2303/commits/700ea5dcd7d955bde4308357502918a1af8f3720.

### Assing a GPU to each process using `CUDA_VISIBLE_DEVICES`

When I use `CUDA_VISIBLE_DEVICES` to assign a GPU to each process, only the `cuda:0` can be seen by the process. So, `size_buffer.cuda(torch.device(rank))` raises no such device error. In this case, all nodes should be set `0` instead of `rank`.

[Full example code](https://gist.github.com/toshihikoyanase/21dc5dd6d899e5173e54f28f41e194bc#file-only_device_zero_is_visible-py)

### Multi-worker setting

When I use multiple GPU nodes, I needed to set the local rank to the device instead of the world rank. Suppose we have two nodes, and each node has two GPUs. The rank will be 0, 1, 2 and 3, but each node has device 0 and device 1. `TorchDistributedTrial(trial, device=local_rank)` resolved this problem.

[Full example code](https://gist.github.com/toshihikoyanase/21dc5dd6d899e5173e54f28f41e194bc#file-multinode-py)

Reply:
Thank you for your detailed investigation!
The above explanation is very helpful for me to understand DDP's behavior!
Adding the `device` argument seems to be a good solution for this issue 👍